bpf: Use "fallthrough;", compile with -Wimplicit-fallthrough

qmonnet opened the pull request:
> **Warning** to backporters:
>
> Only the first commit should be backported.

We can label intentional fall-throughs and tell clang to report unannotated ones, to catch bugs that would be caused by involuntarily falling through between switch labels.

The result is not super aesthetic in builtins.h or jhash.h, but at the same time it's just one keyword so not too much boilerplate either. And it catches bugs. The one fixed in commit 9a641116021a (`"bpf: test: fix pktgen for IPv6 NEXTHDR_DEST option"`), or the first commit in this PR, were found this way.

I wasn't able to pin down the exact version that got support for the relevant attribute, but clang 10 has it for sure and it seems to be much older than that, so we don't have to worry and test for its availability.

Inspired by checkpatch complaining about the `/* fall through */` comments.

qmonnet committed:
bpf: Use "fallthrough;", compile with -Wimplicit-fallthrough

We can label intentional fall-throughs and tell clang to report
unannotated ones, to catch bugs that would be caused by involuntarily
falling through between switch labels.

The bug fixed in commit 9a641116021a ("bpf: test: fix pktgen for IPv6
NEXTHDR_DEST option") was found this way.

I wasn't able to pin down the exact version that got support for the
relevant attribute, but clang 10 has it for sure and it seems to be much
older than that, so we don't have to worry and test for its
availability.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

lmb commented on bpf/lib/egress_policies.h:
```diff
@@ -371,6 +371,7 @@ srv6_create_state_entry(struct __ctx_buff *ctx)
 
 		if (map_update_elem(&SRV6_STATE_MAP6, inner_ips, outer_ips, 0) < 0)
 			return DROP_INVALID;
+		break;
```
These seem like genuine mistakes, maybe split them into a separate commit with a Fixes tag?

qmonnet replied:
Absolutely, I meant to do that in the first place :facepalm:. Thanks! Will update.

lmb reviewed (approved):
Nice!

qmonnet committed:
bpf: Add missing "break;" in srv6_create_state_entry()

Prevent falling through from IPv6 to IPIP when creating SRV6 state
entries based on the encapsulated header protocol.

Caught by compiling with -Wimplicit-fallthrough.

Fixes: bfba74097e81 ("bpf: Handle reply SRv6 traffic")
Signed-off-by: Quentin Monnet <quentin@isovalent.com>
